opentracing/span.go

```diff
@@ -0,0 +1,165 @@
+package opentracing
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// Span is the Appdash implementation of the `opentracing.Span` interface.
+type Span struct {
+	sync.Mutex
+	Recorder      *appdash.Recorder
+	tracer        *Tracer
+	operationName string
+	startTime     time.Time
+	sampled       bool
+	attributes    map[string]string
+	tags          map[string]interface{}
+
+	logs []opentracing.LogData
+}
+
+func newAppdashSpan(operationName string, tracer *Tracer) *Span {
+	return &Span{
+		operationName: operationName,
+		tracer:        tracer,
+		startTime:     time.Now(),
+		logs:          make([]opentracing.LogData, 0),
+	}
+}
+
+// SetOperationName sets the name of the span, overwriting the previous value.
+func (s *Span) SetOperationName(operationName string) opentracing.Span {
+	s.operationName = operationName
+	return s
+}
+
+// Tracer returns the interal opentracing.Tracer
+func (s *Span) Tracer() opentracing.Tracer {
+	return s.tracer
+}
+
+// Finish ends the span.
+//
+// Defers to s.FinishWithOptions() with FinishTime = time.Now()
+func (s *Span) Finish() {
+	s.FinishWithOptions(opentracing.FinishOptions{FinishTime: time.Now()})
+}
+
+// FinishWithOptions finishes the span with opentracing.FinishOptions.
+//
+// Internally, the `appdash.Reporter` reports the span's name, tags,
+// attributes, and log events.
+func (s *Span) FinishWithOptions(opts opentracing.FinishOptions) {
+	if !s.sampled {
+		return
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	s.Recorder.Name(s.operationName) // Set the span's name
+
+	// Convert span tags to annotations.
+	for key, value := range s.tags {
+		val := []byte(fmt.Sprintf("%v", value))
+		s.Recorder.Annotation(appdash.Annotation{Key: key, Value: val})
+	}
+
+	// Record any trace attributes as annotations.
+	for key, value := range s.attributes {
+		s.Recorder.Annotation(appdash.Annotation{Key: key, Value: []byte(value)})
+	}
+
+	// XXX(bg): I'm not too sure how this works in Appdash, but there needs to
+	// be a way to record a key value pair where the value is the payload.
+	for _, log := range s.logs {
+		s.Recorder.LogWithTimestamp(log.Event, log.Timestamp)
+	}
```

Comment:
It's just of type `interface{}`. iirc correctly `LogData.Payload` doesn't have any restrictions/notes on its type and can be any arbitrary type. However, the spec doesn't make any promises of the tracing system using it so we can possibly just not use it.